### PR TITLE
update practitioner

### PR DESCRIPTION
**Pull request by sashrika**
Just an initial version

https://issues.openmrs.org/browse/FM-116

### CODE REVIEW

**Comment by sashrika** [comment]
@harsha89 @surangak  dev completed. Please review.

File: omod/src/main/java/org/openmrs/module/fhir/providers/RestfulPractitionerResourceProvider.java [hunk 1]
```diff
@@ -134,4 +140,35 @@ public MethodOutcome createFHIRPractitioner(@ResourceParam Practitioner practiti
...
+							+ " make sure you have set at lease one non-delete name, Gender and Birthdate to create a new "
+							+ "Person with the given UUID");
+			retVal.setOperationOutcome(outcome);
+			return retVal;
+		}*/
+		outcome.addIssue().setDetails("Person is successfully updated " + p.getUuid());
+		retVal.setOperationOutcome(outcome);
+		return retVal;
+	}
 }
```
Comment (sashrika) [inline omod/src/main/java/org/openmrs/module/fhir/providers/RestfulPractitionerResourceProvider.java]: :octocat:  seems like i have forgotten to update this part

File: api/src/main/java/org/openmrs/module/fhir/api/impl/PractitionerServiceImpl.java [hunk 2]
```diff
@@ -157,16 +158,40 @@ public Practitioner getPractitioner(String id) {
...
+	/**
+	 * @see org.openmrs.module.fhir.api.PractitionerService#updatePractitioner(Practitioner
+	 *      practitioner, IdDt theId)
+	 */
+	public Practitioner updatePractitioner(Practitioner practitioner, IdDt theId) {
+		org.openmrs.api.ProviderService providerService = Context.getProviderService();
+		org.openmrs.Provider retrievedProvider = providerService.getProviderByUuid(theId.getIdPart());
+		if (retrievedProvider != null) { // update existing practitioner
+			retrievedProvider = FHIRPractitionerUtil.updatePractitionerAttributes(practitioner, retrievedProvider);
+			Provider p=Context.getProviderService().saveProvider(retrievedProvider);
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/impl/PractitionerServiceImpl.java]: spaces? and use meaningful variable namess

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java [hunk 3]
```diff
@@ -139,7 +139,39 @@ public static Practitioner generatePractitioner(Provider provider) {
 		return practitioner;
 	}
 	
-	public static Person generateOpenMRSPerson(Practitioner practitioner) {
+	public static Person generateOpenMRSPerson(Person personFromRequest) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java]: Add comments when possible

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java [hunk 4]
```diff
@@ -139,7 +139,39 @@ public static Practitioner generatePractitioner(Provider provider) {
...
+			}
+		}
+		
+		Set<Person> personList = Context.getPersonService().getSimilarPeople(usedName.getFullName(),
+		    1900 + personFromRequest.getBirthdate().getYear(), personFromRequest.getGender()); // filter Persons
+		boolean createPerson = false;
+		if (personList.isEmpty()) {
+			createPerson = true;
+		} else {
+			if (personList.size() != 1) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java]: when there is multiple persons we going to create new oone?

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java [hunk 5]
```diff
@@ -139,7 +139,39 @@ public static Practitioner generatePractitioner(Provider provider) {
...
+				createPerson = true;
+			}
+		}
+		
+		Person personForProvider = null;
+		if (createPerson) {
+			personForProvider = Context.getPersonService().savePerson(personFromRequest);
+		} else {
+			for (Iterator<Person> pers = personList.iterator(); pers.hasNext();) {
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java]: Can be use for(Person person:personList)?

File: api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java [hunk 6]
```diff
@@ -139,7 +139,39 @@ public static Practitioner generatePractitioner(Provider provider) {
...
+				personForProvider = pers.next();
```
Comment (harsha89) [inline api/src/main/java/org/openmrs/module/fhir/api/util/FHIRPractitionerUtil.java]: What is actully ddoing here? Iterating all the loop and person for provider will be the last ?

File: omod/src/main/java/org/openmrs/module/fhir/providers/RestfulPractitionerResourceProvider.java [hunk 7]
```diff
@@ -134,4 +140,35 @@ public MethodOutcome createFHIRPractitioner(@ResourceParam Practitioner practiti
...
+		//p.setBirthdate(new Date());
+		
+		PersonName na = new PersonName();
+		Set<PersonName> set = new TreeSet<PersonName>();
+		na.setGivenName("buruwah");
+		na.setFamilyName("haha");
+		set.add(na);
+		//p.setNames(set);
+		p = Context.getPersonService().savePerson(p);
+		/*try {
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/providers/RestfulPractitionerResourceProvider.java]: remove commented codes

File: omod/src/main/java/org/openmrs/module/fhir/providers/RestfulPractitionerResourceProvider.java [hunk 8]
```diff
@@ -134,4 +140,35 @@ public MethodOutcome createFHIRPractitioner(@ResourceParam Practitioner practiti
...
+	public MethodOutcome updatePractitioner(@ResourceParam Person thePerson, @IdParam IdDt theId) {
+		MethodOutcome retVal = new MethodOutcome();
+		OperationOutcome outcome = new OperationOutcome();
+		org.openmrs.Person p = new org.openmrs.Person();
+		//p.setGender("m");
+		//p.setBirthdate(new Date());
+		
+		PersonName na = new PersonName();
+		Set<PersonName> set = new TreeSet<PersonName>();
+		na.setGivenName("buruwah");
```
Comment (harsha89) [inline omod/src/main/java/org/openmrs/module/fhir/providers/RestfulPractitionerResourceProvider.java]: buruwah? Errgh ugly

**Pull request merged**